src/common/io/io.cpp

```diff
@@ -573,6 +584,53 @@ error::Error FileReader::Rewind() {
 	return error::NoError;
 }
 
+ExpectedSize BufferedReader::Read(vector<uint8_t>::iterator start, vector<uint8_t>::iterator end) {
+	Vsize bytes_read_buffer {0};
+	Vsize bytes_read_wrapper {0};
+
+	if (!rewind_done_) {
+		// Read from the wrapped reader and save copy into the buffer
+		auto bytes_read = wrapped_reader_.Read(start, end);
+		if (!bytes_read) {
+			return bytes_read;
+		}
+		if (!stop_done_) {
+			buffer_.insert(buffer_.end(), start, end);
+		}
+		return bytes_read;
+	}
+
+	// Read from the buffer and/or from the wrapped reader
+	auto ex_bytes_read = buffer_reader_.Read(start, end);
+	if (!ex_bytes_read) {
+		return ex_bytes_read;
+	}
+
+	bytes_read_buffer = ex_bytes_read.value();
+	Vsize iterator_size {static_cast<Vsize>(end - start)};
+	if (iterator_size > bytes_read_buffer) {
+		ex_bytes_read = wrapped_reader_.Read(start + bytes_read_buffer, end);
+		if (!ex_bytes_read) {
+			return ex_bytes_read;
+		}
+		bytes_read_wrapper = ex_bytes_read.value();
+	}
+
+	return bytes_read_buffer + bytes_read_wrapper;
+}
+
+error::Error BufferedReader::Rewind() {
+	if (rewind_done_) {
```

Comment:
If you follow my suggestion above then you don't need this check, but there should be a check for `stop_done_` instead.